src/lib/ares_process.c

```diff
@@ -79,26 +114,29 @@ static void          server_increment_failures(struct server_state *server)
   if (node == NULL) {
     return;
   }
+
   server->consec_failures++;
   ares__slist_node_reinsert(node);
+
+  invoke_server_state_cb(server, ARES_FALSE);
 }
 
 static void server_set_good(struct server_state *server)
 {
   ares__slist_node_t   *node;
   const ares_channel_t *channel = server->channel;
 
-  if (!server->consec_failures) {
-    return;
-  }
-
   node = ares__slist_node_find(channel->servers, server);
   if (node == NULL) {
     return;
   }
 
-  server->consec_failures = 0;
-  ares__slist_node_reinsert(node);
+  if (server->consec_failures > 0) {
+    server->consec_failures = 0;
+    ares__slist_node_reinsert(node);
+  }
+
+  invoke_server_state_cb(server, ARES_TRUE);
```

Comment:
Shouldn't this be inside the above `if()` statement.  I don't see a reason to notify of success on every query.

Reply:
See comment above on the corresponding case for failures.